File: ml-parcial-IIND3118/ml-parcial-IIND3118/temas/B12_svd/python/B12_svd.py
"""
B12. SVD (Singular Value Decomposition)
Tema B12_svd — Introducción al Machine Learning (IIND3118)

CÓMO USARLO
  1. Cambia el bloque marcado con ✏️ abajo por tu dataset.
  2. Corre el archivo:  python B12_svd.py
  3. Las gráficas se guardan en la carpeta figuras/ de este tema.

La explicación del método está en el README.md de esta misma carpeta.
"""

import os, sys, warnings
warnings.filterwarnings("ignore")

import numpy as np
import pandas as pd
import matplotlib
matplotlib.use("Agg")          # no abre ventanas: funciona en cualquier entorno
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split, KFold, cross_val_score, GridSearchCV
from sklearn.preprocessing import scale
from sklearn.linear_model import (LinearRegression, Ridge, Lasso, ElasticNet,
                                  RidgeCV, LassoCV, ElasticNetCV, LogisticRegression)
from sklearn.neighbors import KNeighborsRegressor, KNeighborsClassifier
from sklearn.metrics import (mean_squared_error, mean_absolute_error, r2_score,
                             accuracy_score, confusion_matrix, classification_report,
                             roc_curve, roc_auc_score)
from sklearn.decomposition import PCA, KernelPCA
from sklearn.cluster import KMeans, DBSCAN
from sklearn.ensemble import IsolationForest
from sklearn.neighbors import LocalOutlierFactor
from sklearn import manifold

# ── Guardado automático de figuras ──────────────────────────────────────
AQUI = os.path.dirname(os.path.abspath(__file__))
FIGS = os.path.join(AQUI, "..", "figuras")
os.makedirs(FIGS, exist_ok=True)
_n = [0]
def _save():
    _n[0] += 1
    ruta = os.path.join(FIGS, f"{_n[0]:02d}_figura.png")
    plt.savefig(ruta, dpi=110, bbox_inches="tight"); plt.close()
    print(f"   [figura guardada: figuras/{os.path.basename(ruta)}]")
plt.show = _save                          # cada plt.show() guarda en vez de abrir

RNG = np.random.default_rng(0)

# ╔═════════════════════════════════════════════════════════════════════════╗
# ║  ✏️  CAMBIA SOLO ESTE BLOQUE POR TU DATASET                             ║
# ╚═════════════════════════════════════════════════════════════════════════╝
DATOS  = os.path.join(AQUI, "..", "..", "..", "data", "ejemplo_regresion.csv")
TARGET = "target"          # nombre EXACTO de la columna respuesta. None si no hay.
# ╚═════════════════════════════════════════════════════════════════════════╝

df = pd.read_csv(DATOS).dropna()

if TARGET is not None:
    y_serie = df[TARGET]; X_df = df.drop(columns=[TARGET])
else:
    y_serie = None; X_df = df.copy()

X_df = pd.get_dummies(X_df, drop_first=True).astype(float)
nombres_X = list(X_df.columns)
X  = X_df.values
Xs = scale(X)

if y_serie is not None:
    y = pd.factorize(y_serie)[0] if y_serie.dtype == object else y_serie.values
    ES_CLASIFICACION = (len(np.unique(y)) <= 10) and np.all(y == np.round(y))
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=.30, random_state=0)
    Xs_train, Xs_test, _, _ = train_test_split(Xs, y, test_size=.30, random_state=0)
else:
    y = None; ES_CLASIFICACION = False

def ModeloLineal(**kw):
    return (LogisticRegression(max_iter=5000, **kw) if ES_CLASIFICACION
            else LinearRegression(**kw))

col = y if y is not None else "steelblue"
Z   = PCA().fit_transform(Xs)        # proyección PCA, se usa para graficar
Z2  = Z[:, :2]

print("="*74)
print("B12. SVD (Singular Value Decomposition)")
print("="*74)
print(f"datos: {X.shape[0]} obs. x {X.shape[1]} predictores")
if y is not None:
    print(f"y ('{TARGET}'): " + ("CLASIFICACIÓN" if ES_CLASIFICACION else "REGRESIÓN"))
print()

# ════════════════════════════════════════════════════════════════════════
#  CÓDIGO DEL MÉTODO
# ════════════════════════════════════════════════════════════════════════

# ═══ B12. SVD ═══
# PCA se calcula aquí para poder verificar que SVD == PCA
pca = PCA().fit(Xs)

Xc = Xs - Xs.mean(0)                              # datos centrados
U, S, Vt = np.linalg.svd(Xc, full_matrices=False) # ⬅ False = versión económica

print(f"U: {U.shape}   S: {S.shape}   Vt: {Vt.shape}  (Vt ya está TRANSPUESTA)")
print(f"valores singulares: {np.round(S,3)}")
print(f"\nError de la descomposición: {np.abs(Xc - U@np.diag(S)@Vt).max():.2e}")

# ── SVD == PCA ─────────────────────────────────────────────────────────
print(f"\nvarianzas PCA      : {np.round(pca.explained_variance_,4)}")
print(f"sigma²/(n-1) de SVD: {np.round(S**2/(len(Xs)-1),4)}   <- iguales")

# ── Eckart-Young: el error se sabe en fórmula cerrada ─────────────────
energia = np.cumsum(S**2)/np.sum(S**2)
print(f"\n{'k':>3} | {'energía':>9} | {'error real':>11} | {'error teórico':>13}")
print("-"*45)
for k in range(1, min(6, len(S))+1):
    Ak = U[:,:k]@np.diag(S[:k])@Vt[:k,:]
    print(f"{k:3d} | {energia[k-1]:9.4f} | {np.linalg.norm(Xc-Ak,'fro'):11.4f} | "
          f"{np.sqrt(np.sum(S[k:]**2)):13.4f}")
print("Los valores singulares descartados dicen EXACTAMENTE cuánto se pierde.")

fig, ax = plt.subplots(1, 2, figsize=(12, 4))
ax[0].plot(range(1,len(S)+1), S, "o-"); ax[0].set_yscale("log")
ax[0].set_xlabel("i"); ax[0].set_ylabel("valor singular σᵢ")
ax[0].set_title("Valores singulares (escala log)")
ax[1].plot(range(1,len(S)+1), energia, "o-")
ax[1].axhline(.95, color="red", ls="--", label="95%")
ax[1].set_xlabel("k"); ax[1].set_ylabel("energía acumulada")
ax[1].set_title("Fracción de información retenida"); ax[1].legend()
plt.tight_layout(); plt.show()

print("\nListo. Revisa la carpeta figuras/")

File: ml-parcial-IIND3118/ml-parcial-IIND3118/temas/B12_svd/R/B12_svd.R
# ############################################################################
# B12. SVD
# Tema B12_svd — Introducción al Machine Learning (IIND3118)
#
# CÓMO USARLO
#   1. Cambia el bloque marcado con ✏️ por tu dataset.
#   2. Corre:  Rscript B12_svd.R     (o ábrelo en RStudio)
#   3. Las gráficas se guardan en la carpeta figuras/ de este tema.
#
# La explicación del método está en el README.md de esta misma carpeta.
# ############################################################################

for (p in c("MASS","cluster","dbscan","glmnet","leaps","caret","class","fpc",
            "clusterCrit","isotree","kernlab","Rtsne","psych","car","vegan")) {
  if (!requireNamespace(p, quietly=TRUE))
    try(install.packages(p, repos="https://cloud.r-project.org"), silent=TRUE)
}
suppressMessages({ library(MASS); library(cluster) })

set.seed(0)

# Las gráficas se guardan en figuras/ en vez de abrirse en pantalla
dir.create("../figuras", showWarnings=FALSE)
.n <- 0
guardar <- function() {
  .n <<- .n + 1
  dev.copy(png, sprintf("../figuras/%02d_figura.png", .n), width=900, height=650)
  invisible(dev.off())
}

# ╔═══════════════════════════════════════════════════════════════════════╗
# ║  ✏️  CAMBIA SOLO ESTE BLOQUE POR TU DATASET                           ║
# ╚═══════════════════════════════════════════════════════════════════════╝
DATOS  <- "../../../data/ejemplo_regresion.csv"
TARGET <- "target"        # nombre EXACTO de la columna respuesta. NULL si no hay.
# ╚═══════════════════════════════════════════════════════════════════════╝

df <- na.omit(read.csv(DATOS))

if (!is.null(TARGET)) {
  y_raw <- df[[TARGET]]
  X_df  <- df[, setdiff(names(df), TARGET), drop=FALSE]
} else { y_raw <- NULL; X_df <- df }

X_df <- as.data.frame(model.matrix(~ . - 1, data=X_df))
nombres_X <- names(X_df)
X  <- as.matrix(X_df)
Xs <- scale(X)

if (!is.null(y_raw)) {
  ES_CLASIFICACION <- is.factor(y_raw) || is.character(y_raw) || length(unique(y_raw)) <= 10
  y <- if (ES_CLASIFICACION) as.factor(y_raw) else as.numeric(y_raw)
  idx <- sample(seq_len(nrow(X)), floor(0.70*nrow(X)))
  Xtrain <- X[idx,]; Xtest <- X[-idx,]; ytrain <- y[idx]; ytest <- y[-idx]
  Xs_train <- Xs[idx,]; Xs_test <- Xs[-idx,]
} else { y <- NULL; ES_CLASIFICACION <- FALSE }

pca  <- prcomp(X, center=TRUE, scale.=TRUE)
Z2   <- pca$x[, 1:2]
col_pt <- if (!is.null(y)) as.integer(as.factor(y)) else "steelblue"

cat(strrep("=",74), "\n")
cat("B12. SVD\n")
cat(strrep("=",74), "\n")
cat(sprintf("datos: %d obs. x %d predictores\n\n", nrow(X), ncol(X)))

# ############################################################################
#  CÓDIGO DEL MÉTODO
# ############################################################################

# ============================================================================
# B12. SVD (SINGULAR VALUE DECOMPOSITION)
# ----------------------------------------------------------------------------
# QUÉ HACE: descompone CUALQUIER matriz (no tiene que ser cuadrada ni
# simétrica) en tres piezas:   A = U D V'
# Los VALORES SINGULARES d1 >= d2 >= ... dicen cuánta "información" hay en
# cada dirección. Quedándote con los primeros k obtienes una VERSIÓN
# COMPRIMIDA de la matriz.
#
# ⚠️ ECKART-YOUNG (el teorema clave)
#   Truncar a los primeros k valores singulares da la MEJOR APROXIMACIÓN
#   POSIBLE DE RANGO k. Ninguna otra matriz de rango k se acerca más.
#   Y el error se sabe EN FÓRMULA CERRADA:
#       ||A - A_k||_F = sqrt( suma de los d_i^2 descartados )
#   -> los valores singulares descartados dicen EXACTAMENTE cuánto pierdes.
#
# ⚠️ RELACIÓN CON PCA
#   PCA = SVD SOBRE LOS DATOS CENTRADOS
#     V = los componentes | U*D = los scores | lambda_i = d_i^2/(n-1)
#   Por eso prcomp usa SVD: es MÁS ESTABLE que formar X'X.
#
# APLICACIÓN — LSI (recuperación de información)
#   Matriz término-documento -> SVD -> espacio SEMÁNTICO LATENTE.
#   Resuelve la SINONIMIA: buscar "petroleum" recupera documentos que solo
#   dicen "crude". Ranking por SIMILITUD COSENO (no euclidiana, porque los
#   documentos largos tendrían ventaja injusta).
#
# ⚠️ svd() de R devuelve v.  np.linalg.svd de Python devuelve Vt (TRANSPUESTA).
#    Al traducir, transponer.
#
# CÓMO ELEGIR k: la ENERGÍA ACUMULADA  cumsum(d^2)/sum(d^2)
# ============================================================================

Xc <- scale(Xs, center=TRUE, scale=FALSE)
sv <- svd(Xc)

cat(sprintf("u: %d x %d   d: %d   v: %d x %d\n",
            nrow(sv$u), ncol(sv$u), length(sv$d), nrow(sv$v), ncol(sv$v)))
cat("valores singulares:", round(sv$d, 3), "\n")
cat(sprintf("error de la descomposición: %.2e\n",
            max(abs(Xc - sv$u %*% diag(sv$d) %*% t(sv$v)))))

# SVD == PCA
cat(sprintf("\nd^2/(n-1) == varianzas de PCA: %s\n",
            isTRUE(all.equal(sv$d^2/(nrow(Xc)-1), pca$sdev^2, tolerance=1e-6))))

# Eckart-Young: el error se sabe en fórmula cerrada
energia <- cumsum(sv$d^2)/sum(sv$d^2)
cat(sprintf("\n%3s | %9s | %11s | %13s\n", "k", "energía", "error real", "error teórico"))
for (k in 1:min(6, length(sv$d))) {
  Ak <- sv$u[,1:k,drop=FALSE] %*% diag(sv$d[1:k], nrow=k) %*% t(sv$v)[1:k,,drop=FALSE]
  cat(sprintf("%3d | %9.4f | %11.4f | %13.4f\n", k, energia[k],
              sqrt(sum((Xc-Ak)^2)), sqrt(sum(sv$d[(k+1):length(sv$d)]^2))))
}

par(mfrow=c(1,2))
plot(sv$d, type="b", pch=16, log="y", xlab="i", ylab="valor singular",
     main="Valores singulares (log)")
plot(energia, type="b", pch=16, xlab="k", ylab="energía acumulada",
     main="Información retenida")
abline(h=.95, col="red", lty=2)
par(mfrow=c(1,1))

cat("\nListo. Revisa la carpeta figuras/\n")
